src/com/sun/jna/CallbackReference.java

```diff
@@ -143,31 +144,69 @@ private static Callback getCallback(Class<?> type, Pointer p, boolean direct) {
             throw new IllegalArgumentException("Callback type must be an interface");
         Map<Callback, CallbackReference> map = direct ? directCallbackMap : callbackMap;
         synchronized(pointerCallbackMap) {
-            Callback cb = null;
-            Reference<Callback> ref = pointerCallbackMap.get(p);
-            if (ref != null) {
-                cb = ref.get();
-                if (cb != null && !type.isAssignableFrom(cb.getClass())) {
-                    throw new IllegalStateException("Pointer " + p + " already mapped to " + cb
-                                                    + ".\nNative code may be re-using a default function pointer"
-                                                    + ", in which case you may need to use a common Callback class"
-                                                    + " wherever the function pointer is reused.");
-                }
+            Reference<Callback>[] array = pointerCallbackMap.get(p);
+            Callback cb = getTypeAssignableCallback(type, array);
+            if (cb != null) {
                 return cb;
             }
-            int ctype = AltCallingConvention.class.isAssignableFrom(type)
-                ? Function.ALT_CONVENTION : Function.C_CONVENTION;
-            Map<String, Object> foptions = new HashMap<String, Object>(Native.getLibraryOptions(type));
-            foptions.put(Function.OPTION_INVOKING_METHOD, getCallbackMethod(type));
-            NativeFunctionHandler h = new NativeFunctionHandler(p, ctype, foptions);
-            cb = (Callback)Proxy.newProxyInstance(type.getClassLoader(), new Class[] { type }, h);
+            cb = createCallback(type, p);
+            pointerCallbackMap.put(p, addCallback(array, cb));
+
             // No CallbackReference for this callback
             map.remove(cb);
-            pointerCallbackMap.put(p, new WeakReference<Callback>(cb));
             return cb;
         }
     }
 
+    private static Callback getTypeAssignableCallback(Class<?> type, Reference<Callback>[] array) {
+        if (array != null) {
+            for (int i=0;i < array.length;i++) {
+                Callback cb = array[i].get();
+                if (cb != null && type.isAssignableFrom(cb.getClass())) {
+                    return cb;
+                }
+            }
+        }
+        return null;
+    }
+
+    private static Reference<Callback>[] addCallback(Reference<Callback>[] array, Callback cb) {
```

Comment:
What do you think about switching the signature of this to:

```java
private static void addCallback(Map<Pointer, Reference<Callback>[]> pointerCallbackMap, Pointer pointer, Callback cb)
```

Reply:
I'd rather not change it, as your proposal suggests that it will do more than now: to update pointerCallbackMap. 
That update is in a synchronized block on that object. It is nice to see the related get / put in one location.
If all sync block is extracted as well, then there is an extra map-removal operation in that block that would not belong.
The code for reference:
```
...
        Map<Callback, CallbackReference> map = direct ? directCallbackMap : callbackMap;
        synchronized(pointerCallbackMap) {
            Reference<Callback>[] array = pointerCallbackMap.get(p);
            Callback cb = getTypeAssignableCallback(type, array);
            if (cb != null) {
                return cb;
            }
            cb = createCallback(type, p);
            pointerCallbackMap.put(p, addCallback(array, cb));

            // No CallbackReference for this callback
            map.remove(cb);
            return cb;
        }
...
```

However I'd find a better name for the current addCallback. Perhaps 'expandArrayWithCallback'.